### PR TITLE
show available versions of a chart in catalog

### LINKED ISSUE #350: show available versions of a chart in catalog
API send all versions of charts so we can display them in the catalog : 
image

- [x]  display a list of versions somewhere here : 
- [x]  if an old versions is selected we should show an informationnal message
- [x] we should be able to start any version of a chart

### PR DESCRIPTION
**Pull request by PaulGehin**
Needs #386 before merge

Resolves #350

### CODE REVIEW

**Comment by garronej** [comment]
Salut @PaulGehin, c'est good?

**Comment by PaulGehin** [comment]
Pas convaincu que ce soit bien écrit, mais ça fonctionne ^^

File: src/core/usecases/launcher.ts [hunk 1]
```diff
@@ -533,10 +547,19 @@ export const thunks = {
 
             dispatch(actions.initializationStarted());
 
+            const { availableVersions } =
+                await onyxiaApiClient.getPackageAvailableVersions({
+                    catalogId,
+                    packageName
+                });
+            assert(availableVersions !== undefined && availableVersions.length > 0);
```
Comment (garronej) [inline src/core/usecases/launcher.ts]: We are not in JavaScript.  
We shall never check that the object is actualy of the type it's supose to be.  
We must check the untrusted data once when they enter the system then never again.

File: src/core/usecases/launcher.ts [hunk 2]
```diff
@@ -518,11 +527,16 @@ export const thunks = {
...
+            version: string | undefined;
             formFieldsValueDifferentFromDefault: FormFieldValue[];
         }): ThunkAction =>
         async (...args) => {
-            const { catalogId, packageName, formFieldsValueDifferentFromDefault } =
-                params;
+            const {
+                catalogId,
+                packageName,
+                "version": specificVersion,
```
Comment (garronej) [inline src/core/usecases/launcher.ts]: Should either be named `props_version` (more generic) or `versionSpecifiedAsParam`

File: src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorerCards/CatalogExplorerCards.tsx [hunk 3]
```diff
@@ -21,14 +21,19 @@ export type Props = {
     className?: string;
     packages: {
         packageName: string;
+        version: string;
```
Comment (garronej) [inline src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorerCards/CatalogExplorerCards.tsx]: There is no reason to pass the version number here. It's not displayed anywhere.

File: src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorerCards/CatalogExplorerCards.tsx [hunk 4]
```diff
@@ -21,14 +21,19 @@ export type Props = {
...
         packageHomeUrl?: string;
         catalogId: string;
     }[];
     search: string;
     onSearchChange: (search: string) => void;
-    onRequestLaunch: (params: { packageName: string; catalogId: string }) => void;
+    onRequestLaunch: (params: {
+        packageName: string;
+        catalogId: string;
+        version: string;
```
Comment (garronej) [inline src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorerCards/CatalogExplorerCards.tsx]: Be careful, this is a serious design mistake.  
A package is fully determined by `{packageName, catalogId}` it's the primary key of the table.  
The parent of of `<CatalogExplorerCards />` is able to resolve the version from `{packageName, catalogId}`.

File: src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorer.tsx [hunk 5]
```diff
@@ -76,11 +76,12 @@ export const CatalogExplorer = memo((props: Props) => {
...
     const onRequestLaunch = useConstCallback<
         CatalogExplorerCardsProps["onRequestLaunch"]
-    >(({ packageName, catalogId }) =>
+    >(({ packageName, catalogId, version }) =>
         routes
             .catalogLauncher({
                 catalogId,
-                packageName
+                packageName,
+                version
```
Comment (garronej) [inline src/ui/components/pages/Catalog/CatalogExplorer/CatalogExplorer.tsx]: Resolve the version in filtered package here.

File: src/ui/components/pages/Catalog/CatalogLauncher/CatalogLauncher.tsx [hunk 6]
```diff
@@ -142,6 +144,23 @@ export const CatalogLauncher = memo((props: Props) => {
...
+
+        routes
+            .catalogLauncher({
+                "catalogId": route.params.catalogId,
+                "packageName": route.params.packageName,
+                version,
+                "autoLaunch": route.params.autoLaunch,
+                ...(restorablePackageConfig !== undefined && {
+                    "formFieldsValueDifferentFromDefault":
+                        restorablePackageConfig.formFieldsValueDifferentFromDefault
```
Comment (garronej) [inline src/ui/components/pages/Catalog/CatalogLauncher/CatalogLauncher.tsx]: Damn, I was tired, I remeber writing this with you but it's acually:  

```ts
routes.catalogLauncher({
	...route.params,
	version
})
```

File: src/ui/components/pages/MyServices/MyServicesSavedConfigs/MyServicesSavedConfig/MyServicesSavedConfig.tsx [hunk 7]
```diff
@@ -82,33 +82,44 @@ export const MyServicesSavedConfig = memo((props: Props) => {
                     {friendlyName}
                 </Text>
             </div>
-            <div className={classes.linkAndEditButtonWrapper}>
-                <IconButton
-                    className={classes.linkIcon}
-                    iconId="link"
-                    onClick={onLinkClick}
-                />
+            {editLink.href.includes("version") && (
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServicesSavedConfigs/MyServicesSavedConfig/MyServicesSavedConfig.tsx]: Ok haha,  
Sorry but this is to much of a hack. 😄  
There no ovious good solution here but let's try to keep it under control.

**Comment by garronej** [comment]
Hello Paul,  

Could you:  
1) Address the points I mentionned in the code review.  
2) Drop the commits related to the retrocompatiblity. I'll try to find a compromise that's not too hacky.  

Thanks you for your work,  

Best regards,

**Comment by garronej** [comment]
UP?

**Comment by garronej** [comment]
Done!

**Pull request closed without merging**